polars/polars-core/src/vector_hasher.rs

```diff
@@ -41,24 +43,76 @@ where
         let mut av = AlignedVec::with_capacity(self.len());
 
         self.downcast_iter().for_each(|arr| {
-            av.extend(arr.into_iter().map(|opt_v| {
+            av.extend(arr.values().iter().map(|v| {
                 let mut hasher = random_state.build_hasher();
-                opt_v.hash(&mut hasher);
+                v.hash(&mut hasher);
                 hasher.finish()
-            }))
+            }));
+        });
+
+        // We need a random number that will be our null hash value
+        // just take the memory addresses and hash those
+        let null_h = av.as_ptr() as u64 ^ self as *const Self as u64;
+        let mut hasher = random_state.build_hasher();
+        null_h.hash(&mut hasher);
+        let null_h = hasher.finish();
+
+        let hashes = av.as_mut_slice();
+
+        let mut offset = 0;
+        self.downcast_iter().for_each(|arr| {
```

Comment:
there may be an extra optimization here:

currently: 
```
(values_1, validity_1) -> (hash_1, validity_1) -> (hash with nulls set to constant)_1
(values_2, validity_2) -> (hash_2, validity_2) -> (hash with nulls set to constant)_2
-> combine hashes -> hash
```

```
(values_1, validity_1) -> (hash_1, validity_1)
(values_2, validity_2) -> (hash_2, validity_2)
-> (combine values hashes, AND validities)
-> hash with nulls set to constant
```

this way we perform a single pass to the validity bits instead of a one pass per array.

Reply:
I think that combining the validity bitmaps will lead to more hash duplicates. E.g. the operation is not associative.

I think that in such a caste these rows would lead to a hash collision:

```
[None, 1, 2]
[1, 2, None]
[1, None, 2]
```